clients/python/dd_image_client.py

```diff
@@ -0,0 +1,316 @@
+        img = self._rotate_image_if_needed(img, image_file_path)
+
+        base64_img = self._convert_img_to_base64(img)
+        img.close()
+
+        img_data = [base64_img]
+
+        classes = self._make_image_classification_call(img_data)
+        if classes is not None:
+            print(image_file_path + ': ' + classes['cat'] + ' with probability ' + str(classes['prob']))
+            return classes
+        else:
+            print("error making classification call for " + image_file_path)
+            return None
+
+    def extract_features_from_image_file(self, image_file_path):
+        image_data = []
+        img = Image.open(image_file_path)
+        img = self._rotate_image_if_needed(img, image_file_path)
+
+        base64_img = self._convert_img_to_base64(img)
+        img.close()
+        image_data.append(base64_img)
+
+        response = self._make_feature_extraction_call(image_data)
+
+        if response is not None:
+            print("feature vector for " + image_file_path + " is:")
+            print(response['vals'])
+            return response['vals']
+        else:
+            print ("failed to extract features for " + image_file_path)
+            return None
+
+    def extract_features_from_url(self, url):
+        url_input_list = [url]
+        response = self._make_feature_extraction_call(url_input_list)
+
+        if response is not None:
+            print("vector for " + response['uri'] + " is:")
+            print(response['vals'])
+            return response['vals']
+        else:
+            print("error extracting features for " + url)
+            return None
+
+    def extract_features_from_url_file(self, data_file, target_file):
+        print("processing " + data_file + '\n')
+
+        urls = []
+        url_file = open(data_file, 'r')
+        for url in url_file:
+            urls.append(url.rstrip('\n'))
+
+        print("total number of image URL's to process " + str(len(urls)))
+
+        uri_to_features = {}
+        image_processed = 0
+        num_of_erros = 0
+
+        start_time = time.time()
+        while len(urls) > 0:
+            print("remaining images to process: " + str(len(urls)))
+            if self.batch_size > len(urls):
+                self.batch_size = len(urls)
+
+            urls_to_process = []
+            for i in range(0, self.batch_size):
+                urls_to_process.append(urls.pop())
+
+            t0 = time.time()
+            responses = self._make_feature_extraction_call(urls_to_process)
+
+            if responses is not None:
+                for response in responses:
+                    uri_to_features[response['uri']] = response['vals']
+                else:
+                    num_of_erros += 1
+
+            t1 = time.time()
+            image_processed = image_processed + self.batch_size
+            print("total time it took to process " + str(i) + " images was " + str(t1 - t0) + ' seconds')
+
+        full_file = open(target_file, 'w')
```

Comment:
Very often we may want to extract the features into memory (we do this very often in image similarity search applications). I suggest the extracted features are optionnally dumped to disk.